agent/agent.go

```diff
@@ -75,9 +77,22 @@ func (a *Agent) Commit() (err error) {
 }
 
 func (a *Agent) Push() error {
-	return a.repo.Push(&git.PushOptions{
-		Force: a.Config.Push.Force,
-	})
+	if err := a.repo.Push(&git.PushOptions{}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return err
+	}
+
+	if a.Config.Push.Tags {
+		if err := a.repo.Push(&git.PushOptions{
+			RefSpecs: []config.RefSpec{
+				config.RefSpec("refs/tags/*:refs/tags/*"),
+			},
+			Force: a.Config.Push.Force,
+		}); err != nil && err != git.NoErrAlreadyUpToDate {
+			return err
+		}
+	}
```

Comment:
This should probably be extracted into a separate function for readability purposes. For instance

```go
func (a *Agent) pushTags() (err error} {
	if !a.Config.Push.Tags { return nil }
	err = a.repo.Push(&git.PushOptions{
		RefSpecs: []config.RefSpec{
			config.RefSpec("refs/tags/*:refs/tags/*"),
		},
		Force: a.Config.Push.Force,
	})
	return gitError(err)
}
```

It would also be a good idea to add a 

```go
func gitError(err error) error {
	if err == git.NoErrAlreadyUpToDate { err = nil }
	return err
}
```

and then just

```go
if gitError(err) != nil { return err }
```

Reply:
I'm not a big fan of splitting this much. It make the code less readable IMO due to function jumps.
Tho, the inline error check can be spread out on a separate line if necessary.

Reply:
Could you please do it for me @DataHearth. This is the way I've organised the codebase so far. Functions are split into semantic units where possible. In this case our `push` function does two things: the actual push and a conditional push tags. The push call is simply a call to the library but push tags I think should be separate.

In terms of readability, it actually follows Bob Martin's advice. When you read a function like this:

```go
func (a *Agent) Push() {
	a.repo.Push( ... )
	a.maybePushTags( ... )
}
```

it's actually easier to understand what's going on as you don't have to parse a chunk of code and trace what it does - just read the function names. Them you would only jump into the `pushTags` *if* you are concerned with its implementation. Otherwise, it is enough for you to know that `Push` performs a repo push and then pushes tags.